### PR TITLE
feat: add vault preset

### PR DESCRIPTION
**Pull request by zbindenren**
A preset for [hashicorp vault](https://www.vaultproject.io/).

### CODE REVIEW

**Comment by orlangure** [comment]
Hi @zbindenren and thank you for the contribution!
This looks very cool and has great tests, well done!

There are a couple more steps that we would need to make in order to finalize this feature (see [the contribution guide](https://github.com/orlangure/gnomock/blob/master/CONTRIBUTING.md)): preset registry, swagger docs, and most importantly - github actions and circleci jobs to make sure that vault preset is tested on every build (right now your tests are not executed automatically).

Please let me know if you are available to move forward with the additional steps, or if you would like me to take over and add the missing pieces. 

Thank you!

**Comment by zbindenren** [comment]
Hi @orlangure 

I updated the MR. The swagger tests work too (at least on my laptop). I am not sure if it is good to create a file in `/tmp` in the swagger test, because creating a random file is not possible. I let you decide. Feel free to remove that option:

```json
{
  "preset": {
    "version": "latest",
    "auth_token": "gnomock-vault-token",
    "auth": [
      {
        "path": "k8s_cluster1",
        "type": "kubernetes"
      }
    ],
    "policies": [
      {
        "name": "empty_policy",
        "data": "{}"
      }
    ],
    "token_create": {
      "policies": [
        "default"
      ],
      "file_path": "/tmp/token"
    }
  },
  "options": {
  }
}
```

File: preset/vault/options.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+// WithPolicies configures vault with the provided policies.
+func WithPolicies(policies []Policy) Option {
+	return func(p *P) {
+		p.Policies = policies
+	}
+}
+
+// WithAdditionalToken creates an additional access token with the provided policies and stores it
+// in the provied file path.
```
Comment (orlangure) [inline preset/vault/options.go:37]:
```suggestion
// in the provided file path.
```

File: preset/vault/preset.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+			}
+		}
+
+		for _, pol := range p.Policies {
+			if err := cli.Sys().PutPolicy(pol.Name, pol.Data); err != nil {
+				return fmt.Errorf("failed to write policy %s: %w", pol.Name, err)
+			}
+		}
+
+		if p.TokenCreate != nil {
```
Comment (orlangure) [inline preset/vault/preset.go:120]: I'm not sure this option fits here. Generally preset options modify the container, but this one actually makes changes to the client file system.

I do like the idea of helper functions though. There is [`Ingest`](https://github.com/orlangure/gnomock/blob/97f7ed988cf52e93f58ae19df633512c37842bdb/preset/splunk/init.go#L92) function in Splunk preset for example. It does not modify the client, but it also doesn't belong in the options. What do you think about removing it as an option, but creating a helper function instead that can be called by clients (test code) whenever a new token needs to be configured?

**Review by orlangure — COMMENTED** [review]
Got a closer look at this, amazing job!
I have one concern about `TokenCreate` option, and would like to hear your opinion on it. Otherwise looks great!

File: README.md [hunk 3]
```diff
@@ -122,6 +122,7 @@ Kubernetes | [Go package](https://github.com/orlangure/gnomock/tree/master/prese
 CockroachDB | [Go package](https://github.com/orlangure/gnomock/tree/master/preset/cockroachdb) | [Reference](https://pkg.go.dev/github.com/orlangure/gnomock/preset/cockroachdb?tab=doc) | `v19.2.11`, `v20.1.10` | ✅
 InfluxDB | [Go package](https://github.com/orlangure/gnomock/tree/master/preset/influxdb) | [Reference](https://pkg.go.dev/github.com/orlangure/gnomock/preset/influxdb?tab=doc) | `2.0.4-alpine` | ✅
 Cassandra | [Go package](https://github.com/orlangure/gnomock/tree/master/preset/cassandra) | [Reference](https://pkg.go.dev/github.com/orlangure/gnomock/preset/cassandra?tab=doc) | `4.0`, `3` | ✅
+Vault | [Go package](https://github.com/orlangure/gnomock/tree/master/preset/vault) | [Reference](https://pkg.go.dev/github.com/orlangure/gnomock/preset/vault?tab=doc) | `1.10.11`, `1.11.8`, `1.12.4`, `1.13` | ❌
```
[the remark this answers is not in the document]
Comment (zbindenren) [reply README.md]: Cool, I fixed that

File: preset/vault/preset.go (hunk 2 above)
Comment (zbindenren) [reply preset/vault/preset.go]: I agree with you. I removed the option and created two helper methods which can be handy in tests:

```golang
// Client creates a configured vault client for the provided container and token.
func Client(c *gnomock.Container, token string) (*api.Client, error) {
...
}
```

and 

```golang
// CreateToken creates an additional access token with the provided policies. Use the same password you provided
// with the WithAuthToken option.
func CreateToken(c *gnomock.Container, rootToken string, policies ...string) (string, error) {
...
}
```

If all is ok, I can squash the commits and update the MR if you want.

**Review by orlangure — APPROVED** [review]
This is great, thank you for the contribution!

**Pull request merged**